File: tests/r/test-ecological-data-foundation.R
# test-ecological-data-foundation.R
# testthat tests for ecological-data-foundation skill
# Run: testthat::test_file("tests/r/test-ecological-data-foundation.R")
library(testthat)

DATA_DIR <- file.path(dirname(dirname(rstudioapi::getActiveDocumentContext()$path)), "data")
# Fallback for non-RStudio
if (!dir.exists(DATA_DIR)) DATA_DIR <- file.path("tests", "data")

# ── Helpers ──────────────────────────────────────────────────────────────────
flag_coords <- function(df, lat_col = "decimalLatitude", lon_col = "decimalLongitude") {
  df$QA_status <- "OK"
  lat <- suppressWarnings(as.numeric(df[[lat_col]]))
  lon <- suppressWarnings(as.numeric(df[[lon_col]]))
  df$QA_status[abs(lat) > 90 | abs(lon) > 180] <- "COORD_OUT_OF_RANGE"
  df$QA_status[lat == 0 & lon == 0]             <- "COORD_ZERO"
  df$QA_status[is.na(lat) | is.na(lon)]         <- "MISSING_COORDS"
  df
}

remove_duplicates <- function(df, cols = c("scientificName","decimalLatitude","decimalLongitude","eventDate")) {
  cols_ok <- intersect(cols, names(df))
  df[!duplicated(df[, cols_ok, drop = FALSE]), ]
}

# ── Tests ─────────────────────────────────────────────────────────────────────

test_that("zero coordinates are flagged", {
  df <- data.frame(decimalLatitude = 0.0, decimalLongitude = 0.0, QA_status = "OK",
                   stringsAsFactors = FALSE)
  result <- flag_coords(df)
  expect_equal(result$QA_status[1], "COORD_ZERO")
})

test_that("out-of-range latitude is flagged", {
  df <- data.frame(decimalLatitude = 999.0, decimalLongitude = -50.0, QA_status = "OK",
                   stringsAsFactors = FALSE)
  result <- flag_coords(df)
  expect_equal(result$QA_status[1], "COORD_OUT_OF_RANGE")
})

test_that("valid coordinates pass", {
  df <- data.frame(decimalLatitude = -12.5, decimalLongitude = -55.3, QA_status = "OK",
                   stringsAsFactors = FALSE)
  result <- flag_coords(df)
  expect_equal(result$QA_status[1], "OK")
})

test_that("exact duplicates are removed", {
  df <- data.frame(
    scientificName  = rep("Panthera onca", 3),
    decimalLatitude = rep(-12.0, 3),
    decimalLongitude= rep(-55.0, 3),
    eventDate       = rep("2020-01-01", 3),
    stringsAsFactors = FALSE
  )
  result <- remove_duplicates(df)
  expect_equal(nrow(result), 1)
})

test_that("non-duplicates are retained", {
  df <- data.frame(
    scientificName  = c("Panthera onca", "Puma concolor"),
    decimalLatitude = c(-12.0, -15.0),
    decimalLongitude= c(-55.0, -52.0),
    eventDate       = c("2020-01-01", "2020-06-01"),
    stringsAsFactors = FALSE
  )
  result <- remove_duplicates(df)
  expect_equal(nrow(result), 2)
})

test_that("test dataset has required columns", {
  df <- read.csv(file.path(DATA_DIR, "occurrences_raw.csv"))
  expect_true("decimalLatitude"  %in% names(df))
  expect_true("decimalLongitude" %in% names(df))
  expect_true("scientificName"   %in% names(df))
  expect_true("eventDate"        %in% names(df))
})

test_that("test dataset contains known QA issues", {
  df <- read.csv(file.path(DATA_DIR, "occurrences_raw.csv"), stringsAsFactors = FALSE)
  expect_true("zero_coords"   %in% df$QA_issue)
  expect_true("future_date"   %in% df$QA_issue)
  expect_true("duplicate"     %in% df$QA_issue)
})

test_that("majority of test records are clean", {
  df <- read.csv(file.path(DATA_DIR, "occurrences_raw.csv"), stringsAsFactors = FALSE)
  clean_rate <- mean(df$QA_issue == "none")
  expect_gt(clean_rate, 0.7)
})

test_that("flagging reduces dataset size", {
  df <- read.csv(file.path(DATA_DIR, "occurrences_raw.csv"), stringsAsFactors = FALSE)
  flagged <- flag_coords(df)
  n_clean  <- sum(flagged$QA_status == "OK")
  expect_lt(n_clean, nrow(df))
})

File: tests/python/test_ecological_data_foundation.py
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

"""
Tests for ecological-data-foundation skill scripts.
Run: pytest tests/python/test_ecological_data_foundation.py -v
"""
import pytest
import sys
import math
import csv
import tempfile
import shutil
from pathlib import Path

DATA = Path(__file__).parents[1] / "data"
SCRIPT = Path(__file__).parents[2] / "skills" / "ecological-data-foundation" / "scripts" / "clean_occurrences.py"

# ── import the module functions directly ──────────────────────────────────
sys.path.insert(0, str(SCRIPT.parent))
import importlib.util
spec = importlib.util.spec_from_file_location("clean_occ", SCRIPT)
mod  = importlib.util.module_from_spec(spec)
spec.loader.exec_module(mod)


class TestCoordinateFlags:
    def test_zero_coords_flagged(self):
        import pandas as pd
        df = pd.DataFrame({"decimalLatitude":[0.0],"decimalLongitude":[0.0],"QA_status":["OK"]})
        result = mod.flag_coordinate_issues(df)
        assert result.iloc[0]["QA_status"] == "COORD_ZERO"

    def test_out_of_range_lat_flagged(self):
        import pandas as pd
        df = pd.DataFrame({"decimalLatitude":[999.0],"decimalLongitude":[-50.0],"QA_status":["OK"]})
        result = mod.flag_coordinate_issues(df)
        assert result.iloc[0]["QA_status"] == "COORD_OUT_OF_RANGE"

    def test_out_of_range_lon_flagged(self):
        import pandas as pd
        df = pd.DataFrame({"decimalLatitude":[-10.0],"decimalLongitude":[200.0],"QA_status":["OK"]})
        result = mod.flag_coordinate_issues(df)
        assert result.iloc[0]["QA_status"] == "COORD_OUT_OF_RANGE"

    def test_valid_coords_pass(self):
        import pandas as pd
        df = pd.DataFrame({"decimalLatitude":[-12.5],"decimalLongitude":[-55.3],"QA_status":["OK"]})
        result = mod.flag_coordinate_issues(df)
        assert result.iloc[0]["QA_status"] == "OK"

    def test_missing_coords_flagged(self):
        import pandas as pd
        import numpy as np
        df = pd.DataFrame({"decimalLatitude":[np.nan],"decimalLongitude":[-50.0],"QA_status":["OK"]})
        result = mod.flag_coordinate_issues(df)
        assert result.iloc[0]["QA_status"] == "MISSING_COORDS"


class TestDuplicateRemoval:
    def test_exact_duplicates_removed(self):
        import pandas as pd
        df = pd.DataFrame({
            "scientificName": ["Panthera onca"]*3,
            "decimalLatitude": [-12.0]*3,
            "decimalLongitude": [-55.0]*3,
            "eventDate": ["2020-01-01"]*3,
            "QA_status": ["OK"]*3,
        })
        result = mod.remove_exact_duplicates(df)
        assert len(result) == 1

    def test_non_duplicates_retained(self):
        import pandas as pd
        df = pd.DataFrame({
            "scientificName": ["Panthera onca", "Puma concolor"],
            "decimalLatitude": [-12.0, -15.0],
            "decimalLongitude": [-55.0, -52.0],
            "eventDate": ["2020-01-01", "2020-01-01"],
            "QA_status": ["OK", "OK"],
        })
        result = mod.remove_exact_duplicates(df)
        assert len(result) == 2


class TestTemporalFlags:
    def test_future_date_flagged(self):
        import pandas as pd
        df = pd.DataFrame({"eventDate": ["2099-01-01"], "QA_status": ["OK"]})
        result = mod.check_temporal(df)
        assert result.iloc[0]["QA_status"] == "DATE_FUTURE"

    def test_past_date_passes(self):
        import pandas as pd
        df = pd.DataFrame({"eventDate": ["2019-06-15"], "QA_status": ["OK"]})
        result = mod.check_temporal(df)
        assert result.iloc[0]["QA_status"] == "OK"


class TestEndToEnd:
    def test_pipeline_on_test_dataset(self):
        import pandas as pd
        out = tempfile.mkdtemp()
        try:
            mod.main.__globals__['sys'].argv = ["script", str(DATA/"occurrences_raw.csv"), out]
            # Call functions directly instead
            df = mod.load(str(DATA/"occurrences_raw.csv"))
            df = mod.flag_coordinate_issues(df)
            df = mod.remove_exact_duplicates(df)
            df = mod.check_temporal(df)
            clean   = df[df["QA_status"] == "OK"]
            flagged = df[df["QA_status"] != "OK"]
            # 5 known bad records in test dataset
            assert len(flagged) >= 4, f"Expected ≥4 flagged, got {len(flagged)}"
            assert len(clean) > 70,   f"Expected >70 clean, got {len(clean)}"
        finally:
            shutil.rmtree(out)

    def test_required_cols_validation(self):
        import pandas as pd
        df = pd.DataFrame({"species": ["Panthera onca"], "date": ["2020-01-01"]})
        with pytest.raises((ValueError, KeyError, SystemExit)):
            mod.check_required_cols(df, ["decimalLatitude", "decimalLongitude"])
